### PR TITLE
Interval type improvements

### LINKED ISSUE #430: Handle `INTERVAL` type in the backend
This issue is to ensure that Mathesar can handle the Postgres `INTERVAL` data type.

As part of this issue, we need to ensure that:
- Users can use the API to change a column to `INTERVAL`if it's possible to do so (this should already be complete).
- Users can set and change the `precision` and `fields` of a given `INTERVAL` column via a `type_options` field in the API.
- Automatic type inference during file import suggests `INTERVAL` when it makes sense to do so. 
  - We could use something like https://github.com/oleiade/durations or https://github.com/wroberts/pytimeparse
  - Some work on this is complete, but there are probably improvements to be made in detecting durations.

**Additional Context** 
- #251  (parent issue)
- See PR #368 for code that calculates which types are allowed for a column.
- Inference PRs:
  - #127
  - #135
- See comments on #132 for conversation about inferring durations.

### PR DESCRIPTION
**Pull request by mathemancer**
Fixes #430 

This adds a custom interval type _at the SQLAlchemy level_ that maps to the default PostgreSQL type.  Further, we can now accept `precision` and `fields` arguments when creating or altering columns involving the `INTERVAL` type. 

**Technical details**

The `precision` type option takes an integer (1-6) as input.  The `fields` type option takes a string, and defines which fields the interval stores.  Acceptable strings are:
```
YEAR
MONTH
DAY
HOUR
MINUTE
SECOND
YEAR TO MONTH
DAY TO HOUR
DAY TO MINUTE
DAY TO SECOND
HOUR TO MINUTE
HOUR TO SECOND
MINUTE TO SECOND
```
If both `precision` and `fields` are specified, then `fields` must include seconds, since `precision` applies to the seconds field.

The reason for a custom type was initially to ensure that we didn't convert PostgreSQL `INTERVAL`s into Python `timedelta`s, since that conversion is lossy.  It also standardizes the output and some aspects of input for intervals.  Future PRs will similarly standardize output of other time-related types.

For reference, we'll be using the ISO 8601 spec as reduced by [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339) for standardized output, and always-acceptable (at the DB layer) input.  We will also attempt to handle _any_ string as input using the default PostgreSQL date / time / duration parsing.  So, in the case of intervals, we have strings like
```
f"P{years}Y{months}M{days}DT{hours}H{minutes}M{seconds}S"
```
Each variable is an integer with the exception of seconds for output (seconds can be a float).  For input, decimals are allowed, and will be converted appropriately.  Also, inputs will "carry over" when possible.  Seconds and minutes will aggregate into hours, but hours won't aggregate into days since some days are different numbers of hours around DST changes.  Days won't aggregate into months, but months will aggregate into years.  For output, any missing units (e.g. zero minutes) will have actual zeroes so the client can count on each part being in the returned string.  For input, this is not necessary (but you _do_ need to include the `T` separator between the date and time sections if you include time values).

As a bonus, this PR also fixes some bugs in the constraints API tests that were preventing the pipeline from passing.

Finally, there 

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/filters/operations/apply.py
Comment (kgodey) [inline db/filters/operations/apply.py:1]: What is this empty file for?

**Review by kgodey — APPROVED** [review]
Looks good to me overall, I added a couple of comments. @mathemancer, please resolve before merge.

I also think @silentninja should review this since he more recently worked on date and time types.

File: db/filters/operations/apply.py
Comment (mathemancer) [reply db/filters/operations/apply.py:1]: Oops.  That appears to be cruft from a branch change.  I must have had a file open in a different branch.

File: db/tests/types/test_interval.py
Comment (mathemancer) [reply db/tests/types/test_interval.py:199]: Good idea.

File: db/types/interval.py
Comment (silentninja) [inline db/types/interval.py:58]: I don't have any issues with using `column_expression`,  I am just curious to know if you have thought about setting a local `intervalstyle` instead.

**Review by silentninja — CHANGES_REQUESTED** [review]
Looks good to me. Nice work @mathemancer, I learned a few sqlalchemy tricks, thanks. Can you add the interval standard as `ISO 8601` to the  [Engineering  decision](https://wiki.mathesar.org/en/engineering/decisions)

File: db/types/interval.py
Comment (mathemancer) [reply db/types/interval.py:58]: The problem with that is we need to avoid `psycopg2` interpreting the result as an actual `INTERVAL`.  The column expression casts to `TEXT` at the DB layer, which then gets picked up by `psycopg2` as a python `str`.  The problem with letting `psycopg2` pick up `INTERVAL`s otherwise is that it uses a python `timedelta` to represent them, but `timedelta` makes some different choices (e.g., assuming 30 days can be accumulated into 1 month) and the conversion is lossy.  This would result in erroneous info being passed to the UI, and in fact make viewing some values (e.g., 37 days) impossible.

`intervalstyle` doesn't change the actual return type, and so doesn't really solve the problem by itself.  We'd still need to cast to text using a `column_expression`.  Given that, I opted to handle the formatting ourselves, since the PostgreSQL `iso_8601` style is slightly out-of-spec, and made a few choices that don't work that well for our use case (IMO).  It also makes the formatting that would happen completely visible, rather than hidden in the PostgreSQL docs.

File: mathesar/api/db/viewsets/columns.py
Comment (mathemancer) [reply mathesar/api/db/viewsets/columns.py:72]: Ah, good catch.  I changed that so I could get more debugging output at some point and forgot to revert the change.

**Comment by mathemancer** [comment]
> Looks good to me. Nice work @mathemancer, I learned a few sqlalchemy tricks, thanks. Can you add the interval standard as `ISO 8601` to the [Engineering decision](https://wiki.mathesar.org/en/engineering/decisions)

I'm currently writing up a wiki page that will cover the spec.  It's a little more involved.

File: db/types/interval.py
Comment (silentninja) [reply db/types/interval.py:58]: That makes sense, thanks for the explanation!

**Review by silentninja — APPROVED** [review]

**Pull request merged**
